eth/consensus/clique/snapshot_manager.py

```diff
@@ -0,0 +1,297 @@
+            if self.cast_vote(snapshot, header.coinbase, action):
+                vote = Vote(
+                    signer=signer,
+                    block_number=header.block_number,
+                    subject=header.coinbase,
+                    action=action
+                )
+                snapshot.votes.append(vote)
+
+                tally = snapshot.tallies[header.coinbase]
+                if tally.votes > len(snapshot.signers) / 2:
+                    if tally.action is VoteAction.NOMINATE:
+                        snapshot.signers.append(header.coinbase)
+                        self.logger.debug("New signer added: %s", header.coinbase)
+                    else:
+                        if header.coinbase in snapshot.signers:
+                            snapshot.signers.remove(header.coinbase)
+                            self.logger.debug("Signer removed: %s", header.coinbase)
+
+                    for vote in snapshot.votes.copy():
+                        # Discard any pending votes *from* the added or removed member
+                        if vote.signer == header.coinbase:
+                            if self.retract_vote(snapshot, vote.subject, vote.action):
+                                snapshot.votes.remove(vote)
+
+                        # Discard any pending votes *regarding* the added or removed member
+                        # No need to uncast, the whole tally is going to be removed anyway.
+                        if vote.subject == header.coinbase:
+                            snapshot.votes.remove(vote)
+
+                    snapshot.tallies.pop(header.coinbase)
+
+        self.add_snapshot(snapshot)
+
+        return snapshot.get_frozen_clone()
+
+    def get_or_create_snapshot(self, block_number: int, block_hash: Hash32) -> Snapshot:
+        """
+        Return a snapshot either by creating or retrieving it or raise a ``ValidationError``
+        if the header does not have a known ancestor.
+        """
+        try:
+            snapshot = self.get_snapshot(block_number, block_hash)
+        except SnapshotNotFound:
+            return self.create_snapshot_for(block_hash)
+        else:
+            return snapshot
+
+    def create_snapshot_for(self, block_hash: Hash32) -> Snapshot:
+        """
+        Create a ``Snapshot`` for the given ``block_hash``. This involves traversing
+        backwards through the chain of headers to find a suitable base snapshot either
+        from memory, on disk or by creating it on the fly from a checkpoint header.
+        After we've found a suitable base snapshot, apply all headers from after the
+        base snapshot up to the header of ``block_hash`` to create the requested snapshot.
+        """
+        current_header = header = self._lookup_header(block_hash)
+
+        if is_checkpoint(header.block_number, self._epoch_length):
+            return self._create_snapshot_from_checkpoint_header(current_header)
+
+        parents = []
+        while True:
```

Comment:
Can this become something non-infinite?  Set an upper bound and raise an exception if it exceeds that number of iterations (ideally a value we should never realistically hit).

Reply:
We could add this but let my try arguing against it.

1. There's no chance this code could run infinitely in practice. It either exits earlier or travels up to the next checkpoint header (ultimately that would be genesis) **or** it raises an `ValidationError("Unknown ancestor %s", block_hash)`

2. It's pretty comparable to [other existing code paths](https://github.com/ethereum/py-evm/blob/master/eth/db/header.py#L327-L344) that use that pattern.